src/traces/parcoords/lines.js

```diff
@@ -0,0 +1,405 @@
+    });
+
+    var paletteTexture = regl.texture({
+        shape: [256, 1],
+        format: 'rgba',
+        type: 'uint8',
+        mag: 'nearest',
+        min: 'nearest',
+        data: palette(unitToColor, context, Math.round((context ? contextOpacity : 1) * 255))
+    });
+
+    var glAes = regl({
+
+        profile: false,
+
+        blend: {
+            enable: focusAlphaBlending,
+            func: {
+                srcRGB: 'src alpha',
+                dstRGB: 'one minus src alpha',
+                srcAlpha: 1,
+                dstAlpha: 1 // 'one minus src alpha'
+            },
+            equation: {
+                rgb: 'add',
+                alpha: 'add'
+            },
+            color: [0, 0, 0, 0]
+        },
+
+        depth: {
+            enable: !focusAlphaBlending,
+            mask: true,
+            func: 'less',
+            range: [0, 1]
+        },
+
+        // for polygons
+        cull: {
+            enable: true,
+            face: 'back'
+        },
+
+        scissor: {
+            enable: true,
+            box: {
+                x: regl.prop('scissorX'),
+                y: regl.prop('scissorY'),
+                width: regl.prop('scissorWidth'),
+                height: regl.prop('scissorHeight')
+            }
+        },
+
+        dither: false,
+
+        vert: pick ? pickVertexShaderSource : vertexShaderSource,
+
+        frag: fragmentShaderSource,
+
+        primitive: 'lines',
+        lineWidth: 1,
+        attributes: attributes,
+        uniforms: {
+            resolution: regl.prop('resolution'),
+            viewBoxPosition: regl.prop('viewBoxPosition'),
+            viewBoxSize: regl.prop('viewBoxSize'),
+            dim1A: regl.prop('dim1A'),
+            dim2A: regl.prop('dim2A'),
+            dim1B: regl.prop('dim1B'),
+            dim2B: regl.prop('dim2B'),
+            dim1C: regl.prop('dim1C'),
+            dim2C: regl.prop('dim2C'),
+            dim1D: regl.prop('dim1D'),
+            dim2D: regl.prop('dim2D'),
+            loA: regl.prop('loA'),
+            hiA: regl.prop('hiA'),
+            loB: regl.prop('loB'),
+            hiB: regl.prop('hiB'),
+            loC: regl.prop('loC'),
+            hiC: regl.prop('hiC'),
+            loD: regl.prop('loD'),
+            hiD: regl.prop('hiD'),
+            palette: paletteTexture,
+            colorClamp: regl.prop('colorClamp'),
+            scatter: regl.prop('scatter')
+        },
+        offset: regl.prop('offset'),
+        count: regl.prop('count')
+    });
+
+    var colorClamp = [0, 1];
+
+    function setColorDomain(unitDomain) {
+        colorClamp[0] = unitDomain[0];
+        colorClamp[1] = unitDomain[1];
+    }
+
+    var previousAxisOrder = [];
+
+    function makeItem(i, ii, x, y, panelSizeX, canvasPanelSizeY, crossfilterDimensionIndex, scatter, I, leftmost, rightmost) {
+        var loHi, abcd, d, index;
+        var leftRight = [i, ii];
+        var filterEpsilon = verticalPadding / canvasPanelSizeY;
```

Comment:
It doesn't look like this was ever discussed during the original PR. In order to make it so constraint ranges are honored as precisely as possible I'm planning to remove this epsilon. Or at least make it very small so it only catches floating point rounding errors, though provisionally it doesn't even look like I need to do that, if we're careful about transforming data and constraints identically. @monfera @etpinard any concerns if I do that? It doesn't really matter if you're setting these ranges interactively and just exploring the data, but if you set `constraintrange: [901, 999]` we really should exclude 900 and 1000. Multiselect cannot be as precise of course, I'm just talking about single-range selections here.

Reply:
> any concerns if I do that?

None from me :+1: